uniffi_core/src/ffi/rustfuture.rs

```diff
@@ -116,476 +76,627 @@
-    // `Future`.  Implementing this using `RawWakerVTable` allows us verify that this doesn't
-    // happen.
-    const RAW_WAKER_VTABLE: RawWakerVTable = RawWakerVTable::new(
-        Self::raw_clone,
-        Self::raw_wake,
-        Self::raw_wake_by_ref,
-        Self::raw_drop,
-    );
-
-    /// This function will be called when the `RawWaker` gets cloned, e.g. when
-    /// the `Waker` in which the `RawWaker` is stored gets cloned.
-    unsafe fn raw_clone(self_ptr: *const ()) -> RawWaker {
-        Arc::increment_strong_count(self_ptr as *const Self);
-        RawWaker::new(self_ptr, &Self::RAW_WAKER_VTABLE)
-    }
-
-    /// This function will be called when `wake` is called on the `Waker`. It
-    /// must wake up the task associated with this `RawWaker`.
-    unsafe fn raw_wake(self_ptr: *const ()) {
-        Self::from_raw(self_ptr).wake()
-    }
-
-    /// This function will be called when `wake_by_ref` is called on the
-    /// `Waker`. It must wake up the task associated with this `RawWaker`.
-    unsafe fn raw_wake_by_ref(self_ptr: *const ()) {
-        // This could be optimized by only incrementing the strong count if we end up calling
-        // `schedule_do_wake()`, but it's not clear that's worth the extra complexity
-        Arc::increment_strong_count(self_ptr as *const Self);
-        Self::from_raw(self_ptr).wake()
-    }
-
-    /// This function gets called when a `RawWaker` gets dropped.
-    /// This function gets called when a `RawWaker` gets dropped.
-    unsafe fn raw_drop(self_ptr: *const ()) {
-        drop(Self::from_raw(self_ptr))
+/// RustFuture FFI trait.  This allows `Arc<RustFuture<_, _, _>>` to be cast to
+/// `Arc<dyn RustFutureFFI>`, which is needed to implement the public FFI API.  In particular, this
+/// allows you to use RustFuture functionality without knowing the concrete Future type, which is
+/// unnamable.
+#[doc(hidden)]
+trait RustFutureFFI {
+    fn ffi_poll(self: Arc<Self>, data: *const ());
+    fn ffi_cancel(&self);
+    unsafe fn ffi_complete(&self, return_value: &mut (), call_status: &mut RustCallStatus);
+    fn ffi_free(self: Arc<Self>);
+}
+
+impl<F, T, UT> RustFutureFFI for RustFuture<F, T, UT>
+where
+    // See rust_future_new for an explanation of these trait bounds
+    F: Future<Output = T> + Send + 'static,
+    T: FfiConverter<UT> + Send + 'static,
+    UT: Send + 'static,
+{
+    fn ffi_poll(self: Arc<Self>, data: *const ()) {
+        self.poll(data)
+    }
+
+    fn ffi_cancel(&self) {
+        self.cancel()
+    }
+
+    unsafe fn ffi_complete(&self, return_value: &mut (), call_status: &mut RustCallStatus) {
+        // Unsafely transmute return_value.  This works as long as the foreign code calls the
+        // correct `rust_future_complete_*` function.
+        let return_value = std::mem::transmute::<&mut (), &mut T::ReturnType>(return_value);
+        self.complete(return_value, call_status)
+    }
```

Comment:
I really think `&mut ()` should be `*mut ()` or `NonNull<()>`. `&mut` has aliasing guarantees that we are almost definitely violating. Also, transmute should be the absolute last thing to reach for when other casting methods are not applicable.